### PR TITLE
dev: enums VM error and InternalError

### PR DESCRIPTION
**Pull request by azurwastaken**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #402 

## What is the new behavior?

- Added a new enum that act as Interface for Errors which add the possibility for new errors type in the future without breaking existing code
- Added two enums for InternalError and EVMError (their content must be discussed/modified)

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## More Informations

@Eikix Here is my proposal for your issue about the errors handling, feel free to give me feedback/improvement.

### CODE REVIEW

**Review by Eikix — COMMENTED** [review]
Interesting PR! Let me think about it for a bit,
Summoning @enitrat and @ClementWalter who'll also have an opinion 

I'll check again how Rust handles Error management (i usually use `thiserror`)

File: crates/evm/src/errors.cairo [hunk 1]
```diff
@@ -21,32 +21,57 @@ const WRITE_IN_STATIC_CONTEXT: felt252 = 'KKT: WriteInStaticContext';
 const READ_SYSCALL_FAILED: felt252 = 'KKT: read syscall failed';
 
 #[derive(Drop, Copy, PartialEq)]
-enum EVMError {
+enum EVMErrorEnum {
```
Comment (enitrat) [inline crates/evm/src/errors.cairo:24]: If it's an `enum` then calling it `XEnum` is not a good practice imo. It's already defined as an enum, we wouldn't say `PersonStruct` or `balance_u256` if we have type information

File: crates/evm/src/instructions/block_information.cairo [hunk 2]
```diff
@@ -13,48 +13,48 @@ impl BlockInformation of BlockInformationTrait {
...
+        Result::Err(Errors::EVMError(EVMErrorEnum::NotImplemented))
     }
 
     /// 0x41 - COINBASE
     /// Get the block's beneficiary address.
     /// # Specification: https://www.evm.codes/#41?fork=shanghai
-    fn exec_coinbase(ref self: Machine) -> Result<(), EVMError> {
-        Result::Err(EVMError::NotImplemented)
+    fn exec_coinbase(ref self: Machine) -> Result<(), Errors> {
+        Result::Err(Errors::EVMError(EVMErrorEnum::NotImplemented))
```
Comment (enitrat) [inline crates/evm/src/instructions/block_information.cairo:24]:
```suggestion
        Result::Err(Errors::EVMError::NotImplemented))
```

I'd rather have something like this for internal error as well

```rust
Errors::InternalError::TypeConversionError(TYPE_CONVERSION_ERROR)
```

thought idk if it's possible

**Review by enitrat — COMMENTED** [review]
Few remarks

- This PR only brings one extra layer of abstraction, and I don't feel like there is any benefits to this migration. The logic is still handled the same and the separation between EVM errors and Internal errors doesn't add any value since both end up merged into the `Errors` type which is used in EVM functions

- Overall it makes the code very verbose 

Perhaps a better approach would be to keep it as it currently is, and just implement the an error management logic that acts based on enum variants

**Comment by azurwastaken** [comment]
> Few remarks
> 
> * This PR only brings one extra layer of abstraction, and I don't feel like there is any benefits to this migration. The logic is still handled the same and the separation between EVM errors and Internal errors doesn't add any value since both end up merged into the `Errors` type which is used in EVM functions
> * Overall it makes the code very verbose
> 
> Perhaps a better approach would be to keep it as it currently is, and just implement the an error management logic that acts based on enum variants

I agree with the verbosity, maybe there is a way to shorten the way its called.

To me, those change would help to understand trully the cause of an error and would make the debug process faster. Also, the Errors enum work like a pure virtual interface that allow implementation of new types of errors without having to change a single prototype which is, imo, a good things for sustainability.

**Comment by Eikix** [comment]
Ideally I'd like our error management to be the same as Rust, 
That's not possible for now as the ? operator does not implicitly try to apply the "Into" trait

But ideally I'd like for utils error enum to be decoupled from any other crate

EVMError would also be completely self comprised.

If in evm crate, a function uses a util to compute some information, and that util can return an Result::err, then I'd like for the evm function to perform some conversion in a light way.

How about map_err()?

We could remain rustic while waiting for "?" operator to support our desired behaviour?

**Comment by Eikix** [comment]
Closing this for now, we can revisit the error handling, using the knowledge we developed thanks to this PR and our discussion

**Pull request closed without merging**